crates/matrix-sdk/src/room/mod.rs

```diff
@@ -90,6 +90,64 @@ pub use self::{
     messages::{Messages, MessagesOptions},
 };
 
+type DeduplicatedRequestMap<Key> = Mutex<BTreeMap<Key, Arc<Mutex<Result<(), ()>>>>>;
+
+/// Handler that properly deduplicates requests to the same endpoint, and will
+/// properly report error upwards in case the concurrent request failed.
+pub(crate) struct DeduplicatedRequestHandler<Key> {
+    inflight: DeduplicatedRequestMap<Key>,
+}
+
+impl<Key> Default for DeduplicatedRequestHandler<Key> {
+    fn default() -> Self {
+        Self { inflight: Default::default() }
+    }
+}
+
+impl<Key: Clone + Ord + std::hash::Hash> DeduplicatedRequestHandler<Key> {
+    async fn run<'a, F: Future<Output = Result<()>> + SendOutsideWasm + 'a>(
```

Comment:
Actually even `-> impl Future` should work: we could return `futures_util::future::Either::Left(async move { mutex.lock().await.map_err(…) })` in the early-return case and `Either::Right(async move { match code.await { … } })` in the other case. In both cases, the sync mutex guard won't be captured into the `async move` block, so the `!Send`-across-await problem does not exist.

Reply:
Unfortunately, we can't because of [this sequence](https://github.com/matrix-org/matrix-rust-sdk/pull/2615/files#diff-474f20e47fdcf60feac3f839a81f82fbb2c1fd0bb406388b0262adb60216ce3bR128-R129).